chain/consensus/mir/state_manager.go

```diff
@@ -0,0 +1,263 @@
+package mir
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"sync"
+
+	"github.com/multiformats/go-multiaddr"
+	"go4.org/sort"
+
+	"github.com/filecoin-project/lotus/chain/consensus/hierarchical"
+	"github.com/filecoin-project/mir/pkg/events"
+	"github.com/filecoin-project/mir/pkg/modules"
+	"github.com/filecoin-project/mir/pkg/pb/commonpb"
+	"github.com/filecoin-project/mir/pkg/pb/eventpb"
+	"github.com/filecoin-project/mir/pkg/pb/requestpb"
+	t "github.com/filecoin-project/mir/pkg/types"
+)
+
+type Messages []byte
+
+type StateManager struct {
+
+	// The current epoch number.
+	currentEpoch t.EpochNr
+
+	// For each epoch number, stores the corresponding membership.
+	memberships    []map[t.NodeID]t.NodeAddress
+	membershipLock sync.Mutex
```

Comment:
I'm quite convinced that we should not be using the `membershipLock`. The accesses to the `memberships` must be sequential, since they form part of the deterministically updated state. In fact, if they are accessed concurrently, there is some problem with the program logic, as it introduces non-determinism to the state machine. I.e., if it doesn't work without a lock, it will not work with a lock either. The good news is that I think that the access actually is sequential, except potentially for computing the next proposer ID, which can easily be made sequential as well, as proposed in another comment.

Reply:
Access is sequential within the Mir. But there is another goroutine - Manager.
Removing the lock leads to this:

```

WARNING: DATA RACE
Read at 0x00c002a6ad50 by goroutine 970:
  github.com/filecoin-project/lotus/chain/consensus/mir.(*StateManager).OrderedValidatorsAddresses()
      /Users/alpha/Projects/eudico/chain/consensus/mir/state_manager.go:206 +0x60
  github.com/filecoin-project/lotus/chain/consensus/mir.(*Manager).GetBlockMiner()
      /Users/alpha/Projects/eudico/chain/consensus/mir/manager.go:245 +0x2c
  github.com/filecoin-project/lotus/chain/consensus/mir.Mine()
      /Users/alpha/Projects/eudico/chain/consensus/mir/mine.go:160 +0x10dc
  github.com/filecoin-project/lotus/chain/consensus/hierarchical/subnet/consensus.Mine()
      /Users/alpha/Projects/eudico/chain/consensus/hierarchical/subnet/consensus/consensus.go:81 +0x184
  github.com/filecoin-project/lotus/chain/consensus/hierarchical/subnet/submgr.(*Subnet).mine.func2()
      /Users/alpha/Projects/eudico/chain/consensus/hierarchical/subnet/submgr/subnet.go:294 +0x9c

Previous write at 0x00c002a6ad50 by goroutine 796:
  github.com/filecoin-project/lotus/chain/consensus/mir.(*StateManager).applyNewEpoch()
      /Users/alpha/Projects/eudico/chain/consensus/mir/state_manager.go:138 +0x314
  github.com/filecoin-project/lotus/chain/consensus/mir.(*StateManager).ApplyEvent()
      /Users/alpha/Projects/eudico/chain/consensus/mir/state_manager.go:74 +0x9c
  github.com/filecoin-project/lotus/chain/consensus/mir.(*StateManager).ApplyEvent-fm()
      <autogenerated>:1 +0x40
  github.com/filecoin-project/mir/pkg/modules.ApplyEventsSequentially()
```

Reply:
Can you elaborate more on how to fix this without locking?

Reply:
Implemented this by another way - https://github.com/filecoin-project/eudico/pull/234/commits/d1bf8110ea7d0d295e33c5726e3e5b6f9a1ad5c5